linera-storage-service/src/server.rs

```diff
@@ -310,181 +329,365 @@ impl StoreProcessor for ServiceStoreServer {
+                        .collect::<Vec<_>>(),
+                );
+                streamer(stream);
+                stream = Vec::new();
+
+                loop {
+                    if MAX_PAYLOAD_SIZE >= value.len() {
+                        stream.append(&mut value);
+                        break;
+                    }
+
+                    stream.append(&mut value.drain(..MAX_PAYLOAD_SIZE).collect::<Vec<_>>());
+                    streamer(stream);
+                    stream = Vec::new();
+                }
+
+                stream
+            };
+
+            for (key, value) in key_values {
+                stream = extend_stream(stream, key.len().to_be_bytes().to_vec());
+                stream = extend_stream(stream, value.len().to_be_bytes().to_vec());
+                stream = extend_stream(stream, key);
+                stream = extend_stream(stream, value);
             }
+
+            streamer(stream);
         };
-        Ok(Response::new(response))
+
+        let _ = std::thread::spawn(spawner);
+
+        Ok(Response::new(UnboundedReceiverStream::new(rx)))
     }
 
-    #[instrument(target = "store_server", skip_all, err, fields(n_statements = ?request.get_ref().statements.len()))]
-    async fn process_write_batch_extended(
+    #[instrument(target = "store_server", skip_all, err)]
+    async fn process_write_batch(
         &self,
-        request: Request<RequestWriteBatchExtended>,
+        request: Request<tonic::Streaming<RequestWriteBatch>>,
     ) -> Result<Response<()>, Status> {
-        let request = request.into_inner();
-        let RequestWriteBatchExtended { statements } = request;
+        let mut incoming_stream = request.into_inner();
         let mut batch = Batch::default();
-        for statement in statements {
-            match statement.operation.unwrap() {
-                Operation::Delete(key) => {
-                    batch.delete_key(key);
-                }
-                Operation::Put(key_value) => {
-                    batch.put_key_value_bytes(key_value.key, key_value.value);
-                }
-                Operation::Append(key_value_append) => {
-                    let mut pending_big_puts = self.pending_big_puts.write().await;
-                    match pending_big_puts.get_mut(&key_value_append.key) {
-                        None => {
-                            pending_big_puts
-                                .insert(key_value_append.key.clone(), key_value_append.value);
+
+        let mut state = Vec::new();
+        let mut key = Vec::new();
+        let mut value = Vec::new();
+        let mut len_bytes = Vec::with_capacity(8);
+        while let Some(RequestWriteBatch { stream: message }) = incoming_stream.message().await? {
+            let mut message = VecDeque::from(message);
+            while !message.is_empty() || state.get(2).is_some_and(|len| 0 == *len) {
+                match state.len() {
+                    0 => {
```

Comment:
Instead of 0, 1, 2, 3, an enum indicating the significance of the choice being made should be used.